cdap-app-fabric/src/main/java/co/cask/cdap/route/store/ZKRouteStore.java

```diff
@@ -0,0 +1,205 @@
+                        new FutureCallback<RouteConfig>() {
+                          @Override
+                          public void onSuccess(RouteConfig result) {
+                            routeConfigMap.put(serviceId, routeConfig);
+                          }
+
+                          @Override
+                          public void onFailure(Throwable t) {
+                            settableFuture.setException(t);
+                          }
+                        });
+    try {
+      settableFuture.get();
+    } catch (ExecutionException | InterruptedException ex) {
+      throw Throwables.propagate(ex);
+    }
+  }
+
+  @Override
+  public void delete(final ProgramId serviceId) throws NotFoundException {
+    OperationFuture<String> future = zkClient.delete(getZKPath(serviceId));
+    final SettableFuture settableFuture = SettableFuture.create();
+    Futures.addCallback(future, new FutureCallback<String>() {
+      @Override
+      public void onSuccess(String result) {
+        routeConfigMap.remove(serviceId);
+      }
+
+      @Override
+      public void onFailure(Throwable t) {
+        settableFuture.setException(t);
+      }
+    }, Threads.SAME_THREAD_EXECUTOR);
+
+    try {
+      settableFuture.get();
+    } catch (ExecutionException | InterruptedException ex) {
+      if (ex.getCause() instanceof KeeperException.NoNodeException) {
+        throw new NotFoundException(String.format("Route Config for Service %s was not found.", serviceId));
+      }
+      throw Throwables.propagate(ex);
+    }
+  }
+
+  @Override
+  public RouteConfig fetch(final ProgramId serviceId) throws NotFoundException {
+    RouteConfig routeConfig = routeConfigMap.get(serviceId);
+    if (routeConfig != null) {
+      return routeConfig;
+    }
+
+    Future<RouteConfig> settableFuture = getAndWatchData(serviceId, new ZKRouteWatcher(serviceId));
+    try {
+      return settableFuture.get();
+    } catch (InterruptedException | ExecutionException e) {
+      if (e.getCause() instanceof KeeperException.NoNodeException) {
+        throw new NotFoundException(String.format("Route Config for Service %s was not found.", serviceId));
+      }
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private static String getZKPath(ProgramId serviceId) {
+    return String.format("/routestore/%s", ServiceDiscoverable.getName(serviceId));
+  }
+
+  static final Codec<RouteConfig> ROUTE_CONFIG_CODEC = new Codec<RouteConfig>() {
+
+    @Override
+    public byte[] encode(RouteConfig object) throws IOException {
+      return Bytes.toBytes(GSON.toJson(object.getRoutes()));
+    }
+
+    @Override
+    public RouteConfig decode(byte[] data) throws IOException {
+      Map<String, Integer> routes = GSON.fromJson(Bytes.toString(data), MAP_STRING_INTEGER_TYPE);
+      return new RouteConfig(routes);
+    }
+  };
+
+  private Future<RouteConfig> getAndWatchData(final ProgramId serviceId, Watcher watcher) {
```

Comment:
@chtyim If the node doesn't exist, should we place a watch for that? Can't we just let the next fetch request to try and get the data?